[CALCITE-6909] ClickHouse dialect should limit the Precision and Scale of the Decimal type to be within 76

xuzifu666 opened the pull request:
As description in JIRA: https://issues.apache.org/jira/browse/CALCITE-6909

NobiGo commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -1364,6 +1364,47 @@ private static String toSql(RelNode root, SqlDialect dialect,
+    final String query3 = "select cast(\"product_id\" as decimal(42,2)) "
+        + "from \"product\" ";
+
+    final String expectedClickHouse = "SELECT CAST(`product_id` AS `Decimal32(2)`)\n"
```
```
select cast(age as decimal(2,1)) from test1; --success 
select cast(age as `Decimal32(2)`) from test1; -- failed.
```

xuzifu666 replied:
@NobiGo In my test this syntax is OK:
select cast(1.32, 'DECIMAL256(2)') from system.numbers limit 1;
1.32

xuzifu666 replied:
Thanks for the case, from the case the ClickHouse Dialect cast system contains a bug，cast(a as type) should change to cast(a, type), WDYT? @NobiGo

xuzifu666 replied:
I had test for it:
select cast(1.32 as DECIMAL256(2)) from system.numbers limit 1;
1.32

this is OK  @NobiGo

NobiGo replied:
Therefore, the generated SQL here is flawed. It should not be "\`Decimal32(2)\`" but "Decimal32(2)" .

xuzifu666 replied:
there is some special point here，would try to resolve it.

xuzifu666 replied:
@NobiGo Addressed, PTAL

xuzifu666 commented:
@NobiGo could you help review this PR when you have a free time？ thanks

mihaibudiu commented:
Is this ready or should the "lgtm" label be removed?

NobiGo commented:
> Is this ready or should the "lgtm" label be removed?

I removed the LGTM label because Jira's description was misleading.

NobiGo commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -1364,6 +1364,47 @@ private static String toSql(RelNode root, SqlDialect dialect,
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-6909">[CALCITE-6909]
+   * ClickHouse Dailect can not support accurate precision</a>.
+   */
+  @Test void testClickHouseCastDecimalPrecision() {
```
The test cases here are ineffective for the changes.

xuzifu666 replied:
Thanks @NobiGo , I had changed the test case，before the pr would error out：
Expected: is "SELECT CAST(`product_id` AS DECIMAL(75, 12))\nFROM `foodmart`.`product`"
but: was "SELECT CAST(`product_id` AS DECIMAL(19, 12))\nFROM `foodmart`.`product`"

xuzifu666 commented:
> > Is this ready or should the "lgtm" label be removed?
> 
> I removed the LGTM label because Jira's description was misleading.

I had changed the Jira description to “ClickHouse Dailect can not support accurate precision”，is it fitable？

xuzifu666 commented:
@NobiGo Had update the pr as JIRA suggestion, PTAL

xuzifu666 commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -9652,6 +9666,40 @@ Sql withPostgresqlModifiedDecimalTypeSystem() {
+    Sql withClickHouseModifiedDecimalTypeSystem() {
+      final ClickHouseSqlDialect clickHouseSqlDialect =
+          new ClickHouseSqlDialect(ClickHouseSqlDialect.DEFAULT_CONTEXT
+              .withDataTypeSystem(
```
Make sense, had changed to reuse ClickHouseSqlDialect.TYPE_SYSTEM here, thanks~ @mihaibudiu

mihaibudiu commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -9652,6 +9666,13 @@ Sql withPostgresqlModifiedDecimalTypeSystem() {
+    Sql withClickHouseModifiedDecimalTypeSystem() {
+      final ClickHouseSqlDialect clickHouseSqlDialect =
+          new ClickHouseSqlDialect(ClickHouseSqlDialect.DEFAULT_CONTEXT
+              .withDataTypeSystem(ClickHouseSqlDialect.TYPE_SYSTEM));
```
doesn't this DEFAULT_CONTEXT already have this TYPE_SYSTEM?

xuzifu666 replied:
Done，had fix here，thanks~

mihaibudiu reviewed (approved):
This looks fine, if it is agreed that this is the actual fix for the problem you describe.
